BIP324: Handshake prerequisites

dhruv opened the pull request:
Depends on #25361 for some constants, and on https://github.com/bitcoin-core/secp256k1/pull/1129 for ellswift integrated XDH but can be reviewed independently. Only the last 5 commits belong to this PR.

This PR adds xonly [ECDH](https://gist.github.com/dhruv/5b1275751bc98f3b64bcafce7876b489#Initial_handshake) and [HKDF](https://gist.github.com/dhruv/5b1275751bc98f3b64bcafce7876b489#keys-and-session-id-derivation) key derivation code for BIP324. Unit, bench and fuzz tests are included.

The dependency tree for BIP324 PRs is [here](https://bip324.com/sections/code-review/).

dhruv committed:
BIP324 Cipher Suite

dhruv committed:
HKDF key derivation from ECDH secret for BIP324

dhruv committed:
Fuzz test for BIP324 key derivation

naumenkogs commented:
Concept ACK.
Moving forward with BIP324 is a good idea, and this code seems to not have any negative/unexpected side-effects.

sipa commented on src/test/fuzz/key.cpp:
```diff
@@ -305,4 +305,20 @@ FUZZ_TARGET_INIT(key, initialize_key)
     }
+
+    {
+        // Test ECDH computation with another key
```
You could do a much stronger test: take two private keys as input fuzz data, compute the corresponding public keys, and compute the ecdh secrets from (sec1,pub2) and (sec2,pub1), and check that they match.

It's cryptographic data (hashes and EC operations), so this is very unlikely to actually uncover any bugs unless they're very apparent, but this way the fuzz test actually verifies the full behavior we desire from these functions.

dhruv replied:
Done!

sipa commented on src/util/bip324.h:
```diff
@@ -0,0 +1,32 @@
+
+    // Takes ownership of the ECDH secret and wipes the ephemeral secret once
+    // the derivation is complete.
+    explicit BIP324Keys(ECDHSecret&& ecdh_secret, const Span<uint8_t>& initiator_hdata, const Span<uint8_t>& responder_hdata);
```
Nit: pass Spans by value.

dhruv replied:
Done.

sipa commented on src/util/bip324.cpp:
```diff
@@ -0,0 +1,36 @@
+// Copyright (c) 2021 The Bitcoin Core developers
```
I'm not sure this belongs in util, it seems pretty specific to Bitcoin.

dhruv replied:
I moved this into `net.{h|cpp}` as that's where it will be used. I didn't particularly want to increase the length of those files, so please let me know if the right thing to do is to have it in `src/bip324.{h|cpp}`

sipa commented on src/util/bip324.h:
```diff
@@ -0,0 +1,32 @@
+
+constexpr size_t BIP324_KEY_LEN = 32;
+
+class BIP324Keys
```
Not sure if a class for this is desirable in the first place. These keys shouldn't be long-lived in any case (they're only used to initialize the stream ciphers, and then destroyed). That can just as well be done with a simple function call?

dhruv replied:
Done.

dhruv commented on src/key.cpp:
```diff
@@ -332,6 +333,23 @@ bool CKey::Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const
+    ECDHSecret secret;
+    secret.resize(ECDH_SECRET_SIZE);
+
+    if (secp256k1_ecdh(secp256k1_context_sign, secret.data(), &pubkey_internal,
```
Done - still needed a boolean return value due to the other failure case above.

dhruv commented:
Addressed review comments from @sipa. Ready for further review.

naumenkogs commented on src/net.cpp:
```diff
@@ -399,6 +401,38 @@ static CAddress GetBindAddress(SOCKET sock)
     return addr_bind;
 }
 
+bool derive_bip324_keys(ECDHSecret&& ecdh_secret, const Span<uint8_t> initiator_hdata, const Span<uint8_t> responder_hdata, BIP324Keys& derived_keys)
```
Why snake case for function name `derive_bip324_keys` here?

dhruv replied:
Oh oops. I don't know why I did that. Updated.

naumenkogs commented on src/net.cpp:
```diff
@@ -399,6 +401,38 @@ static CAddress GetBindAddress(SOCKET sock)
+        return false;
+    }
+
+    std::string salt{"bitcoin_v2_shared_secret"};
```
The BIP currently refers to `BitcoinSharedSecret`. I guess to ack this, i'd have to see a new bip :)

dhruv replied:
:) I am working on re-structuring it. It's close. Will keep you updated here.

dhruv commented:
Addressed comments; Rebased with master to incorporate #23413. I am still investigating CI failures which seem like a timeout on Cirrus CI but re-running hasn't helped. Meanwhile, ready for further review.

dhruv commented:
Rebased against master. Ready for further review.

PastaPastaPasta commented on src/bench/ecdh.cpp:
```diff
@@ -0,0 +1,30 @@
+// Copyright (c) 2021 The Bitcoin Core developers
```
nit: if you rebase this at some point, might as well update this

dhruv replied:
Done.

PastaPastaPasta reviewed (commented):
concept ACK / light utACK. I did a high level review, and everything makes sense to me.

I do have 1 tiny nit

dhruv commented:
Rebased. Ready for further review.

dhruv commented:
Rebased. `git range-diff e0680bbce ea94adb 2a2cbfc1e`

Ready for further review.

dhruv commented:
Rebased. Ready for further review.

dhruv commented:
Updated to use xonly ECDH as per changes being added to BIP324 before release. Ready for further review.

fanquake commented on configure.ac:
```diff
@@ -1957,7 +1957,7 @@ LIBS_TEMP="$LIBS"
 LIBS="$LIBS_TEMP"
 
-ac_configure_args="${ac_configure_args} --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig"
+ac_configure_args="${ac_configure_args} --disable-shared --with-pic --enable-benchmark=no --enable-module-recovery --enable-module-schnorrsig --enable-module-ecdh -enable-experimental"
```
You can drop `-enable-experimental`. It should no-longer be required for `--enable-module-ecdh`, and note the missing `-` in any case.

dhruv replied:
Thanks. Done.

sipa commented on src/key.cpp:
```diff
@@ -332,6 +334,40 @@ bool CKey::Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const
+
+    secret.resize(ECDH_SECRET_SIZE);
+    hasher.Finalize(secret.data());
+    memory_cleanse(xonly_ecdh.data(), ECDH_SECRET_SIZE);
```
This is unnecessary, as it's already using secure_allocator, which wipes memory on deallocation.

That said, using a vector with secure_allocator is overkill here; you can just use an array (which doesn't need allocation) - and then you do need the cleanse call.

dhruv replied:
Done - simpler with array + cleanse.

sipa commented on src/test/net_tests.cpp:
```diff
@@ -914,4 +916,70 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+    static const std::string responder_hdata = "4c469c70ba242ae0fc98d4eff6258cf19ecab96611c9c736356a4cf11d66edfa4d2970e56744a6d071861a4cbe2730eb7733a38b166e3df73450ef37112dd32f";
+
+    auto initiator_hdata_vec = ParseHex(initiator_hdata);
+    Span<uint8_t> initiator_hdata_span{initiator_hdata_vec.data(), initiator_hdata_vec.size()};
```
These spans are unnecessary. You can just pass `initiator_hdata_vec` etc. whereever `initiator_hdata_span` is used (they get converted automatically).

dhruv replied:
Done.

sipa commented on src/test/net_tests.cpp:
```diff
@@ -914,4 +916,70 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.session_id.size());
+    BOOST_CHECK_EQUAL(initiator_keys.session_id.size(), responder_keys.session_id.size());
+    BOOST_CHECK_EQUAL(0, memcmp(initiator_keys.session_id.data(), responder_keys.session_id.data(), initiator_keys.session_id.size()));
```
You can use `BOOST_CHECK(initiator_keys.session_id == responder_keys.session_id)` instead here.

dhruv replied:
Done.

sipa commented on src/key.h:
```diff
@@ -22,6 +25,14 @@
+constexpr static size_t ECDH_SECRET_SIZE = CSHA256::OUTPUT_SIZE;
+
+// SHA256("secp256k1_ellsq_xonly_ecdh") proven with a unit test (xonly_ecdh_tag)
+const static auto BIP324_ECDH_TAG = ParseHex("4dbebba5394890bf7370e9ca2ca6112941b974c56950e5c4fef7f00a2bbf3e55");
```
You can also use `const CHashWriter HASHER_BIP324_ECDH = TaggedHash("secp256k1_ellsq_xonly_ecdh");`, and then do the rest of the hashing using `(HashWriter(HASHER_BIP324_ECDH) << [rest of stuff] << [to hash]).GetSHA256()`.

dhruv replied:
Super nice change - much cleaner. Thanks for showing me that. Done.

sipa commented on src/key.cpp:
```diff
@@ -332,6 +334,40 @@ bool CKey::Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const
+    return 1;
+}
+
+bool CKey::ComputeBIP324ECDHSecret(const CPubKey& pubkey, const Span<uint8_t> initiator_hdata, const Span<uint8_t> responder_hdata,
```
You can make the arguments here take `Span<const uint8_t>` - or even `Span<const std::byte>`. I don't think this function needs to be able to mutate the passed in hdata.

dhruv commented:
Addressed review comments from @fanquake and @sipa. Ready for further review.

stratospher commented on src/test/key_tests.cpp:
```diff
@@ -344,4 +345,33 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+    CKey responder_key = DecodeSecret(strSecret2C);
+
+    ECDHSecret initiator_secret, responder_secret;
+    auto initiator_hdata_vec = ParseHex("2deb41da6887640dda029ae41c9c9958881d0bb8e28f6bb9039ee9b7bb11091d62f4cbe65cc418df7aefd738f4d3e96c66365b4d38eefd0a883be64112f4495");
```
to match the [64 bytes encoding](https://github.com/bitcoin/bitcoin/blob/c9db50bc2f8674913e785384c0e323c2df71754d/src/test/net_tests.cpp#L926) in commit c9db50b.
```suggestion
    auto initiator_hdata_vec = ParseHex("2deb41da6887640dda029ae41c9c9958881d0bb8e28f6bb9039ee9b7bb11091d62f4cbe65cc418df7aefd738f4d3e926c66365b4d38eefd0a883be64112f4495");
```

dhruv replied:
Thanks! Done.

dhruv committed:
RFC8439 nonce and counter for ChaCha20

theStack reviewed (commented):
Concept ACK

nit: Considering that `ECDHSecret` always has a fixed size, do we really need to use a `std::vector`? An alternative would be to use `std::array` and remove the `.resize()` calls. The following diff on the top-commit seems to work fine:
```diff
diff --git a/src/key.cpp b/src/key.cpp
index aab8bbda1..791c4c878 100644
--- a/src/key.cpp
+++ b/src/key.cpp
@@ -357,7 +357,6 @@ bool CKey::ComputeBIP324ECDHSecret(const CPubKey& pubkey, const Span<const uint8
     hasher << initiator_hdata << responder_hdata << xonly_ecdh;
     auto secret_uint256 = hasher.GetSHA256();

-    secret.resize(ECDH_SECRET_SIZE);
     memcpy(secret.data(), &secret_uint256, CSHA256::OUTPUT_SIZE);
     memory_cleanse(xonly_ecdh, ECDH_SECRET_SIZE);
     return true;
diff --git a/src/key.h b/src/key.h
index 0fdaef1eb..60f6cce3f 100644
--- a/src/key.h
+++ b/src/key.h
@@ -29,7 +29,7 @@ constexpr static size_t ECDH_SECRET_SIZE = CSHA256::OUTPUT_SIZE;
 const CHashWriter HASHER_BIP324_ECDH = TaggedHash("secp256k1_ellsq_xonly_ecdh");

 // Used to represent a ECDH secret (ECDH_SECRET_SIZE bytes)
-using ECDHSecret = std::vector<uint8_t, secure_allocator<uint8_t>>;
+using ECDHSecret = std::array<uint8_t, ECDH_SECRET_SIZE>;

 /** An encapsulated private key. */
 class CKey
diff --git a/src/test/fuzz/net.cpp b/src/test/fuzz/net.cpp
index 09dac0f17..2ca07b7d7 100644
--- a/src/test/fuzz/net.cpp
+++ b/src/test/fuzz/net.cpp
@@ -89,7 +89,6 @@ FUZZ_TARGET_INIT(bip324, initialize_chainparams)
     FuzzedDataProvider fuzzed_data_provider{buffer.data(), buffer.size()};

     ECDHSecret ecdh_secret;
-    ecdh_secret.resize(ECDH_SECRET_SIZE);
     auto ecdh_secret_bytes = fuzzed_data_provider.ConsumeBytes<uint8_t>(ECDH_SECRET_SIZE);
     ecdh_secret_bytes.resize(ECDH_SECRET_SIZE);

```
Not sure though what security implications this has without the `secure_allocator` being involved.

dhruv committed:
RFC8439 implementation and tests

dhruv committed:
Adding forward secure FSChaCha20

dhruv commented:
This push:
- Rebased
- Addressed comment from @theStack - there's cleansing in the right place and tests for it so we can indeed use `std::array` instead of a secure vector.
- Addressed comment from @stratospher 
- Added a couple more derived secrets per new spec in the working group

Ready for further review.

dhruv commented on src/key.cpp:
```diff
@@ -332,6 +334,34 @@ bool CKey::Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const
+
+    uint8_t xonly_ecdh[ECDH_SECRET_SIZE];
+    assert(secp256k1_ecdh(secp256k1_context_sign, xonly_ecdh, &pubkey_internal,
+                          keydata.data(), bip324_ecdh_hash, nullptr));
```
Thanks for teaching me this. Done.

dhruv commented on src/key.h:
```diff
@@ -156,6 +166,10 @@ class CKey
+
+    // Returns false if an invalid public key is provided
+    bool ComputeBIP324ECDHSecret(const CPubKey& pubkey, const Span<const uint8_t> initiator_hdata, const Span<const uint8_t> responder_hdata,
+                                 ECDHSecret& secret) const;
```
Done.

dhruv commented on src/test/key_tests.cpp:
```diff
@@ -344,4 +345,33 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+    BOOST_CHECK(initiator_key.ComputeBIP324ECDHSecret(responder_key.GetPubKey(), initiator_hdata, responder_hdata, initiator_secret));
+    BOOST_CHECK(responder_key.ComputeBIP324ECDHSecret(initiator_key.GetPubKey(), initiator_hdata, responder_hdata, responder_secret));
+    BOOST_CHECK_EQUAL(initiator_secret.size(), ECDH_SECRET_SIZE);
+    BOOST_CHECK_EQUAL(responder_secret.size(), ECDH_SECRET_SIZE);
```
Done.

dhruv commented on src/test/key_tests.cpp:
```diff
@@ -344,4 +345,33 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+    BOOST_CHECK(responder_key.ComputeBIP324ECDHSecret(initiator_key.GetPubKey(), initiator_hdata, responder_hdata, responder_secret));
+    BOOST_CHECK_EQUAL(initiator_secret.size(), ECDH_SECRET_SIZE);
+    BOOST_CHECK_EQUAL(responder_secret.size(), ECDH_SECRET_SIZE);
+    BOOST_CHECK_EQUAL(0, memcmp(initiator_secret.data(), responder_secret.data(), ECDH_SECRET_SIZE));
```
Done.

dhruv commented on src/net.cpp:
```diff
@@ -438,6 +440,42 @@ static CAddress GetBindAddress(const Sock& sock)
+    if (ecdh_secret.size() != ECDH_SECRET_SIZE) {
+        return false;
+    }
+
```
Done. This also let me optimize the interface to the function as it always succeeds now.

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -917,4 +919,73 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(ECDH_SECRET_SIZE, initiator_secret.size());
+    BOOST_CHECK_EQUAL(ECDH_SECRET_SIZE, responder_secret.size());
+    BOOST_CHECK_EQUAL(0, memcmp(initiator_secret.data(), responder_secret.data(), ECDH_SECRET_SIZE));
```
Done.

dhruv commented:
Latest push:
- Rebased
- Addressed comments form @theStack
- The how HKDF derivation no longer needs the elligator swift bytes from the handshake as those are committed to in the initial ECDH.

dhruv committed:
Enable ECDH computation on secp256k1 keys

dhruv committed:
Bench test for ECDH

dhruv committed:
Fuzz test for ECDH

dhruv commented:
Latest push:
- This PR now depends on [integrated XDH](https://github.com/bitcoin-core/secp256k1/pull/1129) rather than the ECDH module in libsecp256k1

Ready for further review.

sipa commented on src/bench/bip324_ecdh.cpp:
```diff
@@ -0,0 +1,54 @@
+
+    std::array<unsigned char, 32> rnd32;
+    GetRandBytes(rnd32);
+    secp256k1_ellswift_encode(ctx, ell64, &pubkey_internal, rnd32.data());
```
The new ellswift API has a function `secp256k1_ellswift_create` which goes directly from private key + randomness to ellswift encoding, bypassing the need for encoding the public key in the traditional form.

dhruv replied:
Much nicer. Thank you for the reminder. Done.

dhruv commented:
Addressed comments. Ready for further review.

dhruv commented:
Added rekeying salt to key derivation. Ready for further review.

dhruv committed:
Allow for RFC8439 AD in cipher suite interface

dhruv commented:
Addressed comments from @stratospher and a clang-tidy issue. Ready for further review.

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_GARBAGE_TERMINATOR_LEN, initiator_keys.garbage_terminator.size());
+    BOOST_CHECK(initiator_keys.garbage_terminator == responder_keys.garbage_terminator);
+    BOOST_CHECK_EQUAL("118878010356d064", HexStr(Span{initiator_keys.garbage_terminator}));
```
Done.

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.rekey_salt.size());
+    BOOST_CHECK(initiator_keys.rekey_salt == responder_keys.rekey_salt);
+    BOOST_CHECK_EQUAL("060d26411cdedabda638b1a807d0b68cbbb326385881409fcd2f5705091c5842", HexStr(initiator_keys.rekey_salt));
```
Done

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.session_id.size());
+    BOOST_CHECK(initiator_keys.session_id == responder_keys.session_id);
+    BOOST_CHECK_EQUAL("e6b4d42bc1fbce790cf7b419d0128c7caf319b491acfd9fd874f29ca3843fe40", HexStr(initiator_keys.session_id));
```
Done

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.responder_P.size());
+    BOOST_CHECK(initiator_keys.responder_P == responder_keys.responder_P);
+    BOOST_CHECK_EQUAL("82c8d144f1dc7b86993c248db020aecfaed37a5c215bea5a7f1bf4399a85ecca", HexStr(initiator_keys.responder_P));
```
Done

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.responder_L.size());
+    BOOST_CHECK(initiator_keys.responder_L == responder_keys.responder_L);
+    BOOST_CHECK_EQUAL("f31764965c22b792a5ac3f35573b01db27a8d479677c53c43d71bbc2141e8886", HexStr(initiator_keys.responder_L));
```
Done

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.initiator_P.size());
+    BOOST_CHECK(initiator_keys.initiator_P == responder_keys.initiator_P);
+    BOOST_CHECK_EQUAL("bc6617bd1473017b1b40e927c0330b74d1bdd30594a1daa49d89c831f03775cd", HexStr(initiator_keys.initiator_P));
```
Done

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_KEY_LEN, initiator_keys.initiator_L.size());
+    BOOST_CHECK(initiator_keys.initiator_L == responder_keys.initiator_L);
+    BOOST_CHECK_EQUAL("75fa71b052982a72ecb4a0e2a73697857121e4fca2f06e791a5d3c3383d4bb16", HexStr(initiator_keys.initiator_L));
```
Done

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,68 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+    BOOST_CHECK(responder_secret.has_value());
+    BOOST_CHECK(initiator_secret.value() == responder_secret.value());
+    BOOST_CHECK_EQUAL("6cbb9cd446f28623979b16affc4b57c4af95db832cb1a4ff8bc41e3192a33bd8", HexStr(initiator_secret.value()));
+    BOOST_CHECK_EQUAL("6cbb9cd446f28623979b16affc4b57c4af95db832cb1a4ff8bc41e3192a33bd8", HexStr(responder_secret.value()));
```
Done

dhruv commented on src/key.h:
```diff
@@ -22,6 +28,12 @@
 typedef std::vector<unsigned char, secure_allocator<unsigned char> > CPrivKey;
 
+constexpr static size_t ECDH_SECRET_SIZE = CSHA256::OUTPUT_SIZE;
+const auto HASHER_BIP324_ECDH = TaggedHash("secp256k1_ellsq_xonly_ecdh");
```
Done

dhruv commented on src/test/key_tests.cpp:
```diff
@@ -344,4 +346,29 @@ BOOST_AUTO_TEST_CASE(bip340_test_vectors)
+    BOOST_CHECK(responder_secret.has_value());
+    BOOST_CHECK(initiator_secret.value() == responder_secret.value());
+    BOOST_CHECK_EQUAL("6cbb9cd446f28623979b16affc4b57c4af95db832cb1a4ff8bc41e3192a33bd8", HexStr(initiator_secret.value()));
+    BOOST_CHECK_EQUAL("6cbb9cd446f28623979b16affc4b57c4af95db832cb1a4ff8bc41e3192a33bd8", HexStr(responder_secret.value()));
```
Done

stratospher commented on src/key.cpp:
```diff
@@ -332,6 +334,40 @@ bool CKey::Derive(CKey& keyChild, ChainCode &ccChild, unsigned int nChild, const
+                                                        const Span<const std::byte> our_ellswift,
+                                                        bool initiating) const
+{
+    uint8_t xonly_ecdh[ECDH_SECRET_SIZE];
```
ab46788: uint8_t can be made std::byte.

dhruv replied:
@stratospher Upon further thought after our conversation, I think unsigned char makes sense here as that's exactly what all the functions it is used with require. So I'm going with that.

dhruv commented:
Latest push:
- Rebased
- Brought in upstream updates
- Addressed comments from @stratospher 
- Changed ECDH tag for better domain separation
- Rekey salt is now 23 bytes instead of 32 bytes to allow rekeying in a single SHA256 compression

Windows tests failing - but should get resolved once secp/#1129 is merged.

dhruv commented on src/test/net_tests.cpp:
```diff
@@ -906,4 +908,64 @@ BOOST_AUTO_TEST_CASE(initial_advertise_from_version_message)
+
+    BOOST_CHECK_EQUAL(BIP324_GARBAGE_TERMINATOR_LEN, initiator_keys.garbage_terminator.size());
+    BOOST_CHECK(initiator_keys.garbage_terminator == responder_keys.garbage_terminator);
+    BOOST_CHECK_EQUAL("73d7412ff3b42a01", HexStr(Span{initiator_keys.garbage_terminator}));
```
Fixed.

dhruv commented:
Latest push:
- garbage terminator length is now 16 bytes instead of 8 bytes
- bidirectional garbage  terminators in derivation in accordance with the recently release BIP draft

dhruv commented:
Rebased.

dhruv commented:
Latest push:
- Bringing in upstream optimizations to https://github.com/bitcoin-core/secp256k1/pull/1129

dhruv commented:
Bringing in upstream rebase (#26153)

dhruv commented:
Latest push:
- Rebased upstream changes
- Upstream changes include a new rekey ratchet mechanism (forward security) based on ChaCha20 instead of SHA256. This also means we no longer need the rekey salt in the session derivation process. BIP changes are WIP and coming soon.

dhruv commented:
Rebased

dhruv commented:
Rebased

dhruv commented:
Rebased changes from #26153

dhruv commented:
Rebased. Brought in upstream changes.

dhruv commented:
Rebased.

dhruv commented:
Rebased. Brought in changes from https://github.com/bitcoin-core/secp256k1/pull/1129.

dhruv committed:
Merge branch 'bip324-cipher-suite' into bip324-handshake

dhruv committed:
Merge commit '984b3d0a1faabc1657e0dd33432fdadfd78335a9' into bip324-handshake

dhruv committed:
Squashed 'src/secp256k1/' changes from bdf39000b9..8034c67a48

8034c67a48 Add doc/ellswift.md with ElligatorSwift explanation
e90aa4e62e Add ellswift testing to CI
131faedd8a Add ElligatorSwift ctime tests
198a04c058 Add tests for ElligatorSwift
9984bfe476 Add ElligatorSwift benchmarks
f053da3ab7 Add ellswift module implementing ElligatorSwift
76c64be237 Add functions to test if X coordinate is valid
aff948fca2 Add benchmark for key generation
5ed9314d6d Add exhaustive tests for ecmult_const_xonly
b69fe88d5e Add x-only ecmult_const version for x=n/d
427bc3cdcf Merge bitcoin-core/secp256k1#1236: Update comment for secp256k1_modinv32_inv256
647f0a5cb1 Update comment for secp256k1_modinv32_inv256
5658209459 Merge bitcoin-core/secp256k1#1228: release cleanup: bump version after 0.3.0
28e63f7ea7 release cleanup: bump version after 0.3.0

git-subtree-dir: src/secp256k1
git-subtree-split: 8034c67a48dc1334bc74ee4ba239111a23d9789e

dhruv commented:
Rebased.

fanquake commented:
Closing this for now, as it's partly included in #27479.
